File: problems/0621-task-scheduler/typescript/solution.ts
// Task Scheduler — https://leetcode.com/problems/task-scheduler/
export function leastInterval(tasks: string[], n: number): number {
  const counts = new Map<string, number>();
  for (const t of tasks) counts.set(t, (counts.get(t) ?? 0) + 1);

  let maxFreq = 0;
  for (const c of counts.values()) maxFreq = Math.max(maxFreq, c);

  let maxCount = 0;
  for (const c of counts.values()) if (c === maxFreq) maxCount++;

  return Math.max(tasks.length, (maxFreq - 1) * (n + 1) + maxCount);
}

File: problems/0621-task-scheduler/python/solution.py
def leastInterval(tasks, n):
    """Task Scheduler — https://leetcode.com/problems/task-scheduler/"""
    counts = {}
    for t in tasks:
        counts[t] = counts.get(t, 0) + 1
    max_freq = max(counts.values())
    max_count = sum(1 for c in counts.values() if c == max_freq)
    return max(len(tasks), (max_freq - 1) * (n + 1) + max_count)
